### PR TITLE
Allow contracts to deploy

### PR DESCRIPTION
**Pull request by swetharepakula**
Fixes https://github.com/nimakaviani/ethereum-container-broker/issues/23

### CODE REVIEW

**Comment by swetharepakula** [comment]
I can make a commit later to bump the submodule after this is reviewed and merged

File: app/controllers/v2/service_bindings_controller.rb [hunk 1]
```diff
@@ -38,7 +38,10 @@ def update
         Rails.logger.info("got account: #{account}")
         cmd = "solc --combined-json abi,bin /tmp/simple.sol > /tmp/simple.bin"
         output = `#{cmd}`
-        cmd = "node /var/vcap/packages/cf-containers-broker/pusher.js -p http://#{host}:#{node_port} -a #{account} -x foo /tmp/simple.bin 2>&1"
+        if !parsedParameters.has_key?("contract_args")
+          parsedParameters["contract_args"] = []
+        end
+        cmd = "node /var/vcap/packages/cf-containers-broker/pusher.js -p http://#{host}:#{node_port} -a #{account} -x foo -c #{parsedParameters["contract_args"]} /tmp/simple.bin 2>&1"
```
Comment (MHBauer) [inline app/controllers/v2/service_bindings_controller.rb]: use parsed args in cmd and make sure something exists for you to pass in.

**Review by MHBauer — APPROVED** [review]

File: pusher.js [hunk 2]
```diff
@@ -13,8 +14,11 @@ const {flags, args} = CLI.parse({
 const address = flags.address
 const password = flags.password
 const provider = flags.provider
+const cArgs = JSON.parse(flags.constructorArgs)
 const contractPath = args.contract
 
+
+
```
Comment (nimakaviani) [inline pusher.js:21]: unnecessary extra space

File: app/controllers/v2/service_bindings_controller.rb (hunk 1 above)
Comment (nimakaviani) [reply app/controllers/v2/service_bindings_controller.rb]: agreed. lets validate the passed arguments and error out early. This assumes that users provide an array of args in a JSON format later to be used by `pusher.js` but we need to validate it and possibly reconstruct the array before passing it through.

**Comment by nimakaviani** [comment]
Had no luck deploying a contract with multiple input args.

I used [Voting.sol](https://gist.githubusercontent.com/maheshmurthy/3da385a42678c3e36a8328cbe47cae5b/raw/82f9d3b884841ab381b3ea86c143281c65dab89b/Voting.sol) which expects `[]byte32` input arguments.

the following bind command fails:
```
cf bind-service poll simple -c '{"contract_url":"https://gist.githubusercontent.com/maheshmurthy/3da385a42678c3e36a8328cbe47cae5b/raw/82f9d3b884841ab381b3ea86c143281c65dab89b/Voting.sol", "contract_args": ["james","jack","tom"] }'
```

what I see in the broker logs is the following:

```
node /var/vcap/packages/cf-containers-broker/pusher.js -p http://10.244.0.143:32771 -a 2327ca0b6f90a0900c117e96efe192385f36509a -x foo -c ["james", "jack", "tom"] /tmp/simple.bin 2>&1
```
notice the spaces here `-c ["james", "jack", "tom"]` which messes up `cli-flags`.

Tried it with a single argument
```
cf bind-service poll simple -c '{"contract_url":"https://gist.githubusercontent.com/maheshmurthy/3da385a42678c3e36a8328cbe47cae5b/raw/82f9d3b884841ab381b3ea86c143281c65dab89b/Voting.sol","contract_args": ["james"] }'
```

and I see this error
```
undefined:1[james] ^SyntaxError: Unexpect
           ^
(eval):1: syntax error, unexpected ':', expecting end-of-input
defined:1[james] ^SyntaxError: Unexpected token j in JSON at
                              ^>
```


Finally tried passing the entire array as a `string` like the following (even though it is pretty counter intuitive and ugly:

```
cf bind-service poll simple -c '{"contract_url":"https://gist.githubusercontent.com/maheshmurthy/3da385a42678c3e36a8328cbe47cae5b/raw/82f9d3b884841ab381b3ea86c143281c65dab89b/Voting.sol","contract_args": "[\\\"james\\\",\\\"jack\\\"]" }'
```

`pusher.js` chokes as it receives `cArgs` in the format `[james,jack]` without the quotes and failes to iterate over it.

**Comment by swetharepakula** [comment]
So currently this works for me. However this contract just expects a single argument and doesn't care what it is. I created this contract to work with the old version of the code for our demo.

`cf bind-service vote ethService -c '{"contract_url":"https://raw.githubusercontent.com/swetharepakula/blockhead-broker-demo/master/poll.sol", "contract_args":["123"]}'`

and so does this:
`cf bind-service vote ethService -c '{"contract_url":"https://raw.githubusercontent.com/swetharepakula/blockhead-broker-demo/master/poll.sol", "contract_args":[123]}'`

So at least the last error occurs because of the space between the `:` and the array.

**Comment by swetharepakula** [comment]
New commit fixes the multiple arguments problem.
It does not fix the issue with the space in the cf bind-service command. Still have to look into that one.

**Comment by swetharepakula** [comment]
So dug some more. I no longer see a problem with the space in the `cf bind-service`. We occasionally seem to error out when getTransactionReceipt does not succeed the first time and we don't have a Unhandled Promise Rejection error. or something similar to that.

However just because we get that error from node, it does not mean that the deployment was not successful. The ruby parser afterwards errors out because there is some error text before the json output. I am thinking we may need to separate the error output from stdout instead of having it all come together.

**Comment by nimakaviani** [comment]
LGTM.

I tried merging but seems like I dont have access. getting the following error:

```
ERROR: Permission to jberkhahn/cf-containers-broker.git denied to nimakaviani.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

@jberkhahn  can you please merge this and bump https://github.com/nimakaviani/ethereum-container-broker

Also would be good if we can figure out my access issue.

**Pull request merged**